core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -2,52 +2,90 @@
 
 
 import com.google.common.base.Strings;
+import com.google.common.cache.*;
 import io.ddf.exception.DDFException;
-
 import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 
+import io.ddf.misc.ALoggable;
+import io.ddf.misc.Config;
+import org.apache.commons.lang.exception.ExceptionUtils;
+
+import java.util.concurrent.TimeUnit;
+
 /**
- * Created by huandao on 6/11/15.
  */
-public class DDFCache {
+public class DDFCache extends ALoggable {
+
+  private LoadingCache<UUID, DDF> mDDFCache;
+
+  private DDFManager mDDFManager;
+
+  public DDFCache(DDFManager manager) {
+    mDDFManager = manager;
+    /**
+     * Maximum number of DDFs allowed in cache, DDFs will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFs = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_IN_CACHE));
+
+    /**
+     * The number of seconds that the DDF will live in cache since it last used(meaning inserted or accessed)
+     */
+    Long timeToIdle = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.DDF_TIME_TO_IDLE_SECONDS));
+
+    mLog.info(String.format("Maximum number of ddfs in cache %s", maxNumberOfDDFs));
+    mLog.info(String.format("DDF's time to idle = %s seconds", timeToIdle));
+    mDDFCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFs).expireAfterAccess(timeToIdle, TimeUnit.SECONDS).
+        recordStats().removalListener(new DDFRemovalListener())
+        .build(new CacheLoader<UUID, DDF>() {
+      @Override public DDF load(UUID uuid) throws Exception {
+        try {
+          mLog.info(String.format("restoring ddf %s", uuid));
+          return mDDFManager.restoreDDF(uuid);
+        } catch (Exception e) {
+          mLog.error(String.format("Error restoring DDF %s, error = %s", uuid, ExceptionUtils.getStackTrace(e)));
+          throw new DDFException(String.format("DDF with uuid %s does not exist", uuid));
+        }
+      }
+    });
+  }
+
+  private boolean containsDDF(UUID uuid) {
+    return mDDFCache.asMap().containsKey(uuid);
+  }
 
-  private Map<UUID, DDF> mDDFs = new ConcurrentHashMap<UUID, DDF>();
   private Map<String, UUID> mUris = new ConcurrentHashMap<String, UUID>();
 
   public void addDDF(DDF ddf) throws DDFException {
-    mDDFs.put(ddf.getUUID(), ddf);
+    if(ddf.getUUID() == null) {
+      throw new DDFException("uuid is null");
+    }
+    mDDFCache.put(ddf.getUUID(), ddf);
+  }
+
+  public CacheStats getCacheStats() {
+    return mDDFCache.stats();
   }
 
   public void removeDDF(DDF ddf) throws DDFException {
-    mDDFs.remove(ddf.getUUID());
+    mDDFCache.invalidate(ddf.getUUID());
     if(ddf.getUri() != null) {
       mUris.remove(ddf.getUri());
     }
   }
 
   public DDF[] listDDFs() {
-    return this.mDDFs.values().toArray(new DDF[] {});
+    return mDDFCache.asMap().values().toArray(new DDF[]{});
```

Comment:
I suggest to change this signature to Iterator.
Currently we use listDDFs() in a loop condition:
for(ddf : listDDFS()) { ...} so it's better to return Iterator here.